src/Swarm/Language/Typecheck.hs

```diff
@@ -306,46 +322,54 @@ inferModule s@(Syntax _ t) = (`catchError` addLocToTypeErr s) $ case t of
     -- that binding /after/ (i.e. /within/) the application of @ctx1@.
     withBindings ctx1 $
       maybe id ((`withBinding` Forall [] a) . lvVar) mx $ do
-        Module cmdb ctx2 <- inferModule c2
+        Module c2' ctx2 <- inferModule c2
 
-        -- We don't actually need the result type since we're just going
-        -- to return cmdb, but it's important to ensure it's a command
-        -- type anyway.  Otherwise something like 'move; 3' would be
-        -- accepted with type int.
-        _ <- decomposeCmdTy cmdb
+        -- We don't actually need the result type since we're just
+        -- going to return the entire type, but it's important to
+        -- ensure it's a command type anyway.  Otherwise something
+        -- like 'move; 3' would be accepted with type int.
+        _ <- decomposeCmdTy (c2' ^. sType)
 
         -- Ctx.union is right-biased, so ctx1 `union` ctx2 means later
         -- definitions will shadow previous ones.  Include the binder
         -- (if any) as well, since binders are made available at the top
         -- level, just like definitions. e.g. if the user writes `r <- build {move}`,
         -- then they will be able to refer to r again later.
         let ctxX = maybe Ctx.empty ((`Ctx.singleton` Forall [] a) . lvVar) mx
-        return $ Module cmdb (ctx1 `Ctx.union` ctxX `Ctx.union` ctx2)
+        return $
+          Module
+            (Syntax' l (SBind mx c1' c2') (c2' ^. sType))
+            (ctx1 `Ctx.union` ctxX `Ctx.union` ctx2)
 
   -- In all other cases, there can no longer be any definitions in the
   -- term, so delegate to 'infer'.
   _anyOtherTerm -> trivMod <$> infer s
 
--- | Infer the type of a term which does not contain definitions.
-infer :: Syntax -> Infer UType
+-- | Infer the type of a term which does not contain definitions,
+--   returning a type-annotated term.
+infer :: Syntax -> Infer (Syntax' UType)
 infer s@(Syntax l t) = (`catchError` addLocToTypeErr s) $ case t of
-  TUnit -> return UTyUnit
-  TConst c -> instantiate . toU $ inferConst c
-  TDir _ -> return UTyDir
-  TInt _ -> return UTyInt
-  TAntiInt _ -> return UTyInt
-  TText _ -> return UTyText
-  TAntiText _ -> return UTyText
-  TBool _ -> return UTyBool
-  TRobot _ -> return UTyActor
+  TUnit -> return $ Syntax' l TUnit UTyUnit
```

Comment:
Maybe use `Syntax' l t` to make it clearer where `t` is changed? :slightly_smiling_face: 

Another alternative would be:
```Haskell
return $ s{sTerm = UTyUnit}
```

Reply:
@byorgey how about using that pattern arrow on left side to convert between types? If the conversion is easy, it will be inlined and we will get simple `t` on right side.

(just a thought while commuting, dont take it too seriously)